core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,382 @@
+   */
+  def buildQueryPlans(
+    lookups: Seq[ExpressionLookup],
+    relationLevel: Int = 0
+  ): Seq[QueryPlan] = {
+    // Group lookups by the relation at the current level
+    val nextLookupByRelation = lookups
+      .filter(_.relations.size > relationLevel)
+      .groupBy(_.relations(relationLevel).attributeName().getText)
+
+    // Recursively process each group, incrementing the relation level
+    val nextPlans = nextLookupByRelation.toSeq.flatMap { case (relation, groupedLookups) =>
+      buildQueryPlans(groupedLookups, relationLevel + 1).map { plan =>
+        plan.copy(relationChain = relation :: plan.relationChain)
+      }
+    }
+
+    // Collect lookups at the current level (those with no more relations)
+    val currentLevelLookups = lookups.filter(_.relations.size == relationLevel)
+    val currentPlan = if (currentLevelLookups.nonEmpty) {
+      val expressionMappings = currentLevelLookups.groupBy(_.expression).map { case (expr, exprLookups) =>
+        expr -> exprLookups.flatMap(_.attributeName).toSet
+      }
+      Seq(QueryPlan(Nil, expressionMappings))
+    } else {
+      Seq.empty
+    }
+
+    currentPlan ++ nextPlans
+  }
+
+  /**
+   * Executes a single query plan to retrieve entity records and extract the required attributes.
+   * Processes the attained attributes into ExpressionAndResult tuples that can be consumed
+   * by the InputExpressionReassembler.
+   *
+   * @param workspaceId The UUID of the workspace containing the entities to query
+   * @param entityType The type of the starting entity (e.g., "sample_set", "sample")
+   * @param entityName The name of the specific entity to start the query from
+   * @param rootEntityType The expected entity type for the final results
+   * @param plan The QueryPlan containing:
+   *             - relationChain: List of relation names to traverse (e.g., ["samples", "participants"])
+   *             - expressionMappings: Map from expression strings to the set of attribute names
+   *               that expression needs from the entities at the end of the relation chain
+   * @return A Future containing a sequence of ExpressionAndResult tuples, where each tuple contains:
+   *         - The original expression string as the key
+   *         - A Map from entity name to Try[Iterable[AttributeValue] containing the extracted
+   *           attribute values for that expression
+   */
+  def executeQueryPlan(workspaceId: UUID,
+                       entityType: String,
+                       entityName: String,
+                       rootEntityType: String,
+                       plan: QueryPlan,
+                       entityLookups: Seq[ExpressionLookup] = Seq.empty
+  )(implicit
+    executionContext: ExecutionContext
+  ): Future[Seq[ExpressionAndResult]] =
+    repository.dataSource
+      .inTransaction { _ =>
+        repository.queries.queryRelatedRecordsWithArray(workspaceId, entityType, entityName, plan.relationChain)
+      }
```

Comment:
if each QueryPlan is executed in its own transaction, there exists the possibility that the data will change between subsequent plans and yield some strange results. Would it be better to have a single overarching transaction?